packages/loopring_v3/contracts/core/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -201,72 +201,93 @@ library ExchangeBlocks
+            assembly {
+                block_auxiliaryData := add(blockAuxData, 64)
+            }
+
             require(
-                _block.auxiliaryData.length == header.numConditionalTransactions,
+                block_auxiliaryData.length == header.numConditionalTransactions,
                 "AUXILIARYDATA_INVALID_LENGTH"
             );
 
             // Run over all conditional transactions
             uint minTxIndex = 0;
-            for (uint i = 0; i < _block.auxiliaryData.length; i++) {
-                ExchangeData.AuxiliaryData memory auxiliaryData = _block.auxiliaryData[i];
+            bytes memory txData = new bytes(ExchangeData.TX_DATA_AVAILABILITY_SIZE);
+            for (uint i = 0; i < block_auxiliaryData.length; i++) {
+                // Load the data from auxiliaryData, which is still encoded as calldata
+                uint txIndex;
+                bool approved;
+                bytes memory auxData;
+                assembly {
+                    // Offset to block_auxiliaryData[i]
+                    let auxOffset := mload(add(block_auxiliaryData, add(32, mul(32, i))))
+                    // Load `txIndex` (pos 0) and `approved` (pos 1) in block_auxiliaryData[i]
+                    txIndex := mload(add(add(32, block_auxiliaryData), auxOffset))
+                    approved := mload(add(add(64, block_auxiliaryData), auxOffset))
+                    // Load `data` (pos 2)
+                    let auxDataOffset := mload(add(add(96, block_auxiliaryData), auxOffset))
+                    auxData := add(add(32, block_auxiliaryData), add(auxOffset, auxDataOffset))
+                }
+
                 // Each conditional transaction needs to be processed from left to right
-                require(auxiliaryData.txIndex >= minTxIndex, "AUXILIARYDATA_INVALID_ORDER");
+                require(txIndex >= minTxIndex, "AUXILIARYDATA_INVALID_ORDER");
 
-                minTxIndex = auxiliaryData.txIndex + 1;
+                minTxIndex = txIndex + 1;
 
-                if (auxiliaryData.approved) {
+                if (approved) {
                     continue;
                 }
 
                 // Get the transaction data
-                bytes memory txData = _block.readTransactionData(auxiliaryData.txIndex);
+                _block.data.readTransactionData(txIndex, _block.blockSize, txData);
 
                 // Process the transaction
                 ExchangeData.TransactionType txType = ExchangeData.TransactionType(
                     txData.toUint8(0)
                 );
-                uint txDataOffset = 1;
+                uint txDataOffset = 0;
 
                 if (txType == ExchangeData.TransactionType.DEPOSIT) {
                     DepositTransaction.process(
                         S,
                         ctx,
                         txData,
                         txDataOffset,
-                        auxiliaryData.data
+                        auxData
```

Comment:
Brecht, do you mean now we simply pass in a "auxData as bytes", so each conditional transction can decide whether to further decode these data or not?  Previously each tx's aux data will be decoded before being passed into `process`.

What if we don't use one single `bytes` but use something like `bytes[]` for aux data, will it simplify logics between line 218 and 231?

Reply:
> Brecht, do you mean now we simply pass in a "auxData as bytes", so each conditional transction can decide whether to further decode these data or not? Previously each tx's aux data will be decoded before being passed into process.

At this place there isn't that much difference any more, the solidity costs are there just by using complex `ABIEncoderV2` data types heavily in function arguments. If `approved == true` we can skip decoding the `AuxiliaryData.data` for each tx completely.

> What if we don't use one single bytes but use something like bytes[] for aux data, will it simplify logics between line 218 and 231?

Not totally sure, but I can't see that being cheaper and the logic would remain pretty much the same because `bytes[]` is very similar to `ExchangeData.AuxiliaryData[]`. In any case, we want to also avoid decoding the array elements in `LoopringIOExchangeOwner` so it would essentially be the same there for both. Abi encoded `bytes[]` just wraps the data for each array element in an extra `bytes` encoding vs abi encoded `ExchangeData.AuxiliaryData[]` wraps the complete data in a single extra bytes encoding.